azure/scope/managedcontrolplane.go

```diff
@@ -328,7 +328,12 @@ func (s *ManagedControlPlaneScope) IsIPv6Enabled() bool {
 
 // IsVnetManaged returns true if the vnet is managed.
 func (s *ManagedControlPlaneScope) IsVnetManaged() bool {
-	return true
+	isManaged, err := virtualnetworks.New(s).IsManaged(context.TODO())
```

Comment:
instead of making a new service on the fly in here, have you considered putting this directly into the subnets.go `IsManaged` function? That would allow for proper handling of the error and you'd be able to pass in the context instead of using context.TODO.

The downside is that it would also affect self-managed clusters but the current implementation of `IsVnetManaged()` in `cluster.go` uses the value of ID and tags in the Vnet spec part of `AzureCluster.Spec` which  is not ideal since Spec could be modified by a user. It would be better to always check from the source of truth like we're doing here, i.e. the actual Azure vnet resource.

I would also be careful with where `IsVnetManaged` is being called since we are now making an Azure API call every time (more costly than a hardcoded return true :) ) and from glancing at the subnets code I can already see a few redundant uses that could be removed (for example we're embedding it in the SubnetSpec but then never using it [here](https://github.com/kubernetes-sigs/cluster-api-provider-azure/blob/2a711856de20735337d59ee391202cb45d8f49e2/azure/scope/managedcontrolplane.go#L256)). Long term we might even want to cache the result so we don't have to make the same API call multiple time per reconcile loop (GETs are cheap but still...)

Reply:
> have you considered putting this directly into the subnets.go IsManaged function

I looked at that and it's already calling the Scope's `IsVnetManaged` so I took that to mean that the `AzureManagedControlPlane` Scope is the one that needed to be updated since it's working for `AzureCluster`. I'm new enough to the code though that I might be misunderstanding the way the resource Scope(s) work.

> for example we're embedding it in the SubnetSpec but then never using it

I had been wondering about that myself. It led me through a few false starts when trying to fix this :)

> Long term we might even want to cache the result so we don't have to make the same API call multiple time per reconcile loop

Is there anything special to the way it does caching or is it just a new static field on the ManagedControlPlaneScope struct (like the `MachineCache` does)? I'm happy to add that now if you'd like.

Reply:
> I looked at that and it's already calling the Scope's IsVnetManaged so I took that to mean that the AzureManagedControlPlane Scope is the one that needed to be updated since it's working for AzureCluster

totally reasonable, that's what my first instinct would have told me too. I just happen to have the context that `IsVnetManaged` is legacy code and that we've reworked the way to check if a resource is managed since.

> Is there anything special to the way it does caching or is it just a new static field on the ManagedControlPlaneScope struct (like the MachineCache does)? I'm happy to add that now if you'd like.

Nothing special, just a way to save the value on scope just like we do for VM SKUs in MachineCache. If you want to try and add it in this PR go for it!

Reply:
I updated the `subnets.go`'s `IsManaged` however I'm running into an order of operations issue where it's calling a check to the subnet's IsManaged after the Vnet has already been deleted (when not using BYO Vnet). I added a fall back in that scenario to the old logic, which allows it to move on, but I'm not happy about it. I will debug a bit deeper on Monday to find a better way.

Reply:
I tried both approaches (update `IsVnetManaged()` in `azure/scope/managedcontrolplane.go` and update `IsManaged` in `azure/services/subnets/subnets.go`. They both work for this particular use case, but in the `azure/scope/managedcontrolplane.go` scenario we do end up making a bunch of regular GET calls which is not desireable. 

The reason for this seems obvious enough, as the subnets implementation only invokes `IsManaged()` in its `Delete` reconciler.

So functionally I think the current solution solves the bug. My current concern is the following scenario:

- AKS template with a reference to a pre-existing VNET
- no reference to a pre-existing subnet (so we expect a "default" subnet to be created in the pre-existing VNET)

Is that actually a known-working scenario? If so, it seems like relying on the VNET's "managed" status (looking for the owned tag) to stand in for the subnet's "managed" status wouldn't work. We'd actually *want* to clean up such a subnet is this scenario.

If I'm describing something that isn't possible (for example, if it's required that when you BYO VNET you also must BYO subnet), then this lgtm.

Reply:
> If I'm describing something that isn't possible (for example, if it's required that when you BYO VNET you also must BYO subnet), then this lgtm.

Good point. Let me see if there's something I can do for the scenario when you don't BYO subnet. It's definitely harder since you can't tag/label subnets. :/

Reply:
I think this documentation accurately describes the "self-managed" requirements (not necessarily AKS) for vnets and subnets:

https://capz.sigs.k8s.io/topics/custom-vnet.html#pre-existing-vnet-and-subnets

Specifically:

> When providing a vnet, it is required to also provide the two subnets that should be used for control planes and nodes.

We can definitely add an AKS-contextualized version of that for the AKS scenario (for example, there's no "control plane nodes" in an AKS scenario) that matches how BYO subnet actually works with AKS, e.g. from the `az` CLI help text:

```
    --vnet-subnet-id                           : The ID of a subnet in an existing VNet into which
                                                 to deploy the cluster.
```

tl;dr my scenario described above where you bring your own VNET but rely upon capz to create the subnet is not actually a relevant AKS scenario

Reply:
> tl;dr my scenario described above where you bring your own VNET but rely upon capz to create the subnet is not actually a relevant AKS scenario

Is that going to cause any backwards compatibility issues? If not, I won't worry about it and leave it as-is.

Reply:
@CecileRobertMichon and I were discussing this offline (we should probably try not to do that as much as possible!) and she had the idea to simply turn the subnet delete into a no-op as a potentially more elegant solution to this. I've scaffolded up that idea here:

- https://github.com/kubernetes-sigs/cluster-api-provider-azure/pull/2525

What do we think about that?

Reply:
I was wondering something similar when I noticed that the vnet deletion finished and deleted the subnet before the subnet deletion fully reconciled. The approach works for me but I defer to the folks with more experience in the code base. :)

Reply:
Yes, your observation/insight there was key. Not only does the VNET delete its subnets automatically, but in an environment where concurrent reconcilers are in-progress, it means that if we try do delete subnets and VNETs discretely we have to account for the actual Azure dependency graph relationship between the two and either (1) strictly serialize the order of deletions (we aren't and don't want to do this IMO) or (2) add sane error handling to account for the fact that the subnet may have already been deleted by the VNET and thus return successful from `Delete`.

So it is simpler to not delete the subnet discretely, it seems. However, given the architecture of the type system that Azure resources conform to (subnets are something we treat in capz like a first class service resource) it's a bit "weird" that we're no-op'ing during `Delete`; so I'll track that PR and see what folks think about that approach.

I didn't mean by submitting my own PR to prefer *my* code over *yours*, hopefully that didn't come across! Whichever solution lands (and we'll land it soon) it's certainly entirely due to your observation, investigation, and persistence that we can put this bug behind us. Thanks again for finding this gremlin!

Reply:
> I didn't mean by submitting my own PR to prefer my code over yours, hopefully that didn't come across!

Not at all. Team effort! I was just waiting on closing this PR until I was sure the other one was getting merged.

> Thanks again for finding this gremlin!

Glad I could help. :)